test/server_test.cpp

```diff
@@ -0,0 +1,184 @@
+    return msg.has_process() && msg.process().has_word() && 
+        msg.process().has_state() && msg.process().has_turn();   
+}
+
+class MockCoreFixture : public ::testing::Test
+{
+public:
+    MockCoreFixture() : mMockStream(new MockClientReaderWriter<ProviderMsg, ProviderMsg>()),
+        mMockStub(std::make_shared<MockGameCoreStub>()),
+        mStateMachine(std::make_shared<StateMachine<GlobalState>>()),
+        mNetworkInterface(NetworkInterface::Create(mMockStub)),
+        mGameManager(EventLoop::Create(), mStateMachine, mNetworkInterface) {}
+    
+
+    void SetUp() {
+        spdlog::set_level(spdlog::level::err);
+        EXPECT_CALL(*mMockStub, ProviderRaw(_))
+            .Times(AtLeast(1)).WillOnce(Return(mMockStream));
+        mThread = std::make_unique<std::thread>([&] {
+            mGameManager.Start();
+        });
+
+        while (!mNetworkInterface->IsConnected()) {}
+    }
+
+    void TearDown() {
+        mGameManager.Stop();
+        mThread->join();
+    }
+
+    void ProcessMsgFromCore(const ProviderMsg &msg) {
+        EXPECT_CALL(*mMockStream, Read(_)).Times(1).WillOnce(
+            DoAll(SetArgPointee<0>(msg), Return(true)));
+        mNetworkInterface->OnEventHappens(mNetworkInterface->ReceiveMsg());
+
+        while (mGameManager.HasEventToProcess()) {}
+    }
+
+    void RegisterAndGameStart(int roomId, 
+        const std::vector<unsigned int> &userIds) 
+    {
+        EXPECT_CALL(*mMockStream, Write(_, _)).Times(AtLeast(1));
+        mNetworkInterface->SendMsg(
+            *MsgBuilder::CreateRegisterArgs(0, "codename", "CODENAMES", 10, 4));
+        ProcessMsgFromCore(*MsgBuilder::CreateRegisterRet(0, ErrorNumber::OK));
+
+        EXPECT_CALL(*mMockStream, Write(
+            AllOf(NotifyMsgArgsMatcher(ErrorNumber::OK, roomId, userIds[0]),
+                StartStateMsgMatcher()), _)).Times(1);
+
+        EXPECT_CALL(*mMockStream, Write(
+            AllOf(NotifyMsgArgsMatcher(ErrorNumber::OK, roomId, userIds[1]),
+                StartStateMsgMatcher()), _)).Times(1);
+
+        EXPECT_CALL(*mMockStream, Write(
+            AllOf(NotifyMsgArgsMatcher(ErrorNumber::OK, roomId, userIds[2]),
+                StartStateMsgMatcher()), _)).Times(1);
+
+        EXPECT_CALL(*mMockStream, Write(
+            AllOf(NotifyMsgArgsMatcher(ErrorNumber::OK, roomId, userIds[3]),
+                StartStateMsgMatcher()), _)).Times(1);
+        
+        ProcessMsgFromCore(*MsgBuilder::CreateStartGameArgs(0, roomId, userIds,
+            MsgBuilder::CreateStartGameSettings(15)));
+    }
+protected:
+    MockClientReaderWriter<ProviderMsg, ProviderMsg> *mMockStream;
+    std::shared_ptr<MockGameCoreStub> mMockStub;
+    std::shared_ptr<StateMachine<GlobalState>> mStateMachine;
+    std::shared_ptr<NetworkInterface> mNetworkInterface;
+    GameManager<GlobalState> mGameManager;
+    std::unique_ptr<std::thread> mThread;
+};
+
+TEST_F(MockCoreFixture, RegisterAndGameStart) 
+{
+    auto roomId = 1002;
```

Comment:
define it as class member:
mTestRoomId = 1002;
mTestUserId = {};

Reply:
why?

Reply:
the rule of programming.
'don't repeat yourself'

right now you define the same variable in every test case.

Reply:
but if I want to add a test case with 5 or more players, they don't share same value; besides, if I follow your suggestion, let them be class members, it do will save some space though and maybe improve argument passing efficiency of RegisterAnd..() function, but every time before I call RegisterAnd...() I shall reassign their value which introduces another cost. So in a nutshell, I didn't figure out which one is better and why.

Reply:
1. keep you code clean before talking about the future.
2. even if you want to setup more scenarios, you should use classic programming model to reuse/simplify your code. for example,  define different fixtures for these tests,  use inheritance to manage these fixtures. don't just repeat the code. I admit copy/paste is most flexible, but in most cases, it's not neat.

Reply:
well, it's just test code, don't spend too much on such details, it's more reasonable to focus on development of client side now

Reply:
I believe a good style saves the time, not costs more.  Also, this is a practice/learning,  not a serious project.

Anyway, I approved.